File: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/builders/RegularBuilder.java
/*
 * Pixel Dungeon
 * Copyright (C) 2012-2015  Oleg Dolya
 *
 * Shattered Pixel Dungeon
 * Copyright (C) 2014-2017 Evan Debenham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.shatteredpixel.shatteredpixeldungeon.levels.builders;

import com.shatteredpixel.shatteredpixeldungeon.levels.rooms.Room;
import com.shatteredpixel.shatteredpixeldungeon.levels.rooms.connection.ConnectionRoom;
import com.shatteredpixel.shatteredpixeldungeon.levels.rooms.special.ShopRoom;
import com.shatteredpixel.shatteredpixeldungeon.levels.rooms.standard.EntranceRoom;
import com.shatteredpixel.shatteredpixeldungeon.levels.rooms.standard.ExitRoom;
import com.shatteredpixel.shatteredpixeldungeon.levels.rooms.standard.StandardRoom;
import com.watabou.utils.Random;

import java.util.ArrayList;
import java.util.LinkedHashSet;

//Introduces the concept of a major path, and branches
// with tunnels padding rooms placed in them
public abstract class RegularBuilder extends Builder {
	
	// *** Parameter values for level building logic ***
	// note that implementations do not have to use al of these variables
	
	protected float pathVariance = 45f;
	
	public RegularBuilder setPathVariance( float var ){
		pathVariance = var;
		return this;
	}
	
	//path length is the percentage of pathable rooms that are on
	protected float pathLength = 0.5f;
	//The chance weights for extra rooms to be added to the path
	protected float[] pathLenJitterChances = new float[]{0, 1, 0};
	
	public RegularBuilder setPathLength( float len, float[] jitter ){
		pathLength = len;
		pathLenJitterChances = jitter;
		return this;
	}
	
	protected float[] pathTunnelChances = new float[]{1, 3, 1};
	protected float[] branchTunnelChances = new float[]{2, 2, 1};
	
	public RegularBuilder setTunnelLength( float[] path, float[] branch){
		pathTunnelChances = path;
		branchTunnelChances = branch;
		return this;
	}
	
	protected float extraConnectionChance = 0.2f;
	
	public RegularBuilder setExtraConnectionChance( float chance ){
		extraConnectionChance = chance;
		return this;
	}
	
	// *** Room Setup ***
	
	protected Room entrance = null;
	protected Room exit = null;
	protected Room shop = null;
	
	protected ArrayList<Room> multiConnections = new ArrayList<>();
	protected ArrayList<Room> singleConnections = new ArrayList<>();
	
	protected void setupRooms(ArrayList<Room> rooms){
		for(Room r : rooms){
			r.setEmpty();
		}
		
		entrance = exit = shop = null;
		singleConnections.clear();
		multiConnections.clear();
		for (Room r : rooms){
			if (r instanceof EntranceRoom){
				entrance = r;
			} else if (r instanceof ExitRoom) {
				exit = r;
			} else if (r instanceof ShopRoom && r.maxConnections(Room.ALL) == 1){
				shop = r;
			} else if (r.maxConnections(Room.ALL) > 1){
				multiConnections.add(r);
			} else if (r.maxConnections(Room.ALL) == 1){
				singleConnections.add(r);
			}
		}

		//this weights larger rooms to be much more likely to appear in the main loop, by placing them earlier in the multiconnections list
		weightRooms(multiConnections);
		Random.shuffle(multiConnections);
		multiConnections = new ArrayList<>(new LinkedHashSet<>(multiConnections));
	}
	
	// *** Branch Placement ***
	
	protected void weightRooms(ArrayList<Room> rooms){
		for (Room r : rooms.toArray(new Room[0])){
			if (r instanceof StandardRoom){
				for (int i = 1; i < ((StandardRoom) r).sizeCat.connectionWeight(); i++)
					rooms.add(r);
			}
		}
	}
	
	//places the rooms in roomsToBranch into branches from rooms in branchable.
	//note that the three arrays should be separate, they may contain the same rooms however
	protected void createBranches(ArrayList<Room> rooms, ArrayList<Room> branchable,
	                                     ArrayList<Room> roomsToBranch, float[] connChances){
		
		int i = 0;
		float angle;
		int tries;
		Room curr;
		ArrayList<Room> connectingRoomsThisBranch = new ArrayList<>();
		
		float[] connectionChances = connChances.clone();
		while (i < roomsToBranch.size()){
			
			connectingRoomsThisBranch.clear();
			curr = Random.element(branchable);
			
			int connectingRooms = Random.chances(connectionChances);
			if (connectingRooms == -1){
				connectionChances = connChances.clone();
				connectingRooms = Random.chances(connectionChances);
			}
			connectionChances[connectingRooms]--;
			
			for (int j = 0; j < connectingRooms; j++){
				ConnectionRoom t = ConnectionRoom.createRoom();
				tries = 3;
				
				do {
					angle = placeRoom(rooms, curr, t, randomBranchAngle(curr));
					tries--;
				} while (angle == -1 && tries > 0);
				
				if (angle == -1) {
					for (Room r : connectingRoomsThisBranch){
						r.clearConnections();
						rooms.remove(r);
					}
					connectingRoomsThisBranch.clear();
					break;
				} else {
					connectingRoomsThisBranch.add(t);
					rooms.add(t);
				}
				
				curr = t;
			}
			
			if (connectingRoomsThisBranch.size() != connectingRooms){
				continue;
			}
			
			Room r = roomsToBranch.get(i);
			
			tries = 10;
			
			do {
				angle = placeRoom(rooms, curr, r, randomBranchAngle(curr));
				tries--;
			} while (angle == -1 && tries > 0);
			
			if (angle == -1){
				for (Room t : connectingRoomsThisBranch){
					t.clearConnections();
					rooms.remove(t);
				}
				connectingRoomsThisBranch.clear();
				continue;
			}
			
			for (int j = 0; j <connectingRoomsThisBranch.size(); j++){
				if (Random.Int(3) <= 1) branchable.add(connectingRoomsThisBranch.get(j));
			}
			if (r.maxConnections(Room.ALL) > 1 && Random.Int(3) == 0) {
				if (r instanceof StandardRoom){
					for (int j = 0; j < ((StandardRoom) r).sizeCat.connectionWeight(); j++){
						branchable.add(r);
					}
				} else {
					branchable.add(r);
				}
			}
			
			i++;
		}
	}
	
	protected float randomBranchAngle( Room r ){
		return Random.Float(360f);
	}
	
}

File: pixel-dungeon-web/src/levels/builders/RegularBuilder.js
/*
 * Pixel Dungeon Web Port
 * Ported from Shattered Pixel Dungeon
 * Copyright (C) 2012-2015 Oleg Dolya
 * Copyright (C) 2014-2025 Evan Debenham
 *
 * Regular Builder - Base class for regular level builders
 * Source: core/src/main/java/com/shatteredpixel/shatteredpixeldungeon/levels/builders/RegularBuilder.java
 */

import { Builder } from './Builder.js';
import { Random } from '../../utils/Random.js';
import { TunnelRoom } from '../rooms/connection/TunnelRoom.js';

/**
 * Regular Builder
 * Introduces the concept of a main path and branches
 * Source: RegularBuilder.java:35-247
 */
export class RegularBuilder extends Builder {
    constructor() {
        super();

        // Path variance (angle variation)
        // Source: RegularBuilder.java:42
        this.pathVariance = 45;

        // Path length (percentage of rooms on main path)
        // Source: RegularBuilder.java:50-53
        this.pathLength = 0.25;
        this.pathLenJitterChances = [0, 0, 0, 1];

        // Tunnel chances for path and branches
        // Source: RegularBuilder.java:61-62
        this.pathTunnelChances = [2, 2, 1];
        this.branchTunnelChances = [1, 1, 0];

        // Extra connection chance
        // Source: RegularBuilder.java:71
        this.extraConnectionChance = 0.30;

        // Room categories
        // Source: RegularBuilder.java:80-87
        this.entrance = null;
        this.exit = null;
        this.shop = null;

        this.mainPathRooms = [];
        this.multiConnections = [];
        this.singleConnections = [];
    }

    /**
     * Set path variance
     * Source: RegularBuilder.java:44-47
     */
    setPathVariance(variance) {
        this.pathVariance = variance;
        return this;
    }

    /**
     * Set path length
     * Source: RegularBuilder.java:55-59
     */
    setPathLength(len, jitter) {
        this.pathLength = len;
        this.pathLenJitterChances = jitter;
        return this;
    }

    /**
     * Set tunnel length
     * Source: RegularBuilder.java:64-68
     */
    setTunnelLength(path, branch) {
        this.pathTunnelChances = path;
        this.branchTunnelChances = branch;
        return this;
    }

    /**
     * Set extra connection chance
     * Source: RegularBuilder.java:73-76
     */
    setExtraConnectionChance(chance) {
        this.extraConnectionChance = chance;
        return this;
    }

    /**
     * Setup room categories
     * Source: RegularBuilder.java:89-130
     */
    setupRooms(rooms) {
        // Reset all room connections
        // Source: RegularBuilder.java:90-92
        for (const r of rooms) {
            r.setEmpty();
        }

        // Clear categories
        // Source: RegularBuilder.java:94-97
        this.entrance = null;
        this.exit = null;
        this.shop = null;
        this.mainPathRooms = [];
        this.singleConnections = [];
        this.multiConnections = [];

        // Categorize rooms
        // Source: RegularBuilder.java:98-110
        for (const r of rooms) {
            if (r.isEntrance && r.isEntrance()) {
                this.entrance = r;
            } else if (r.isExit && r.isExit()) {
                this.exit = r;
            } else if (r.maxConnections(0) > 1) {  // RoomDirection.ALL = 0
                this.multiConnections.push(r);
            } else if (r.maxConnections(0) === 1) {
                this.singleConnections.push(r);
            }
        }

        // Weight larger rooms to be more likely on main path
        // Source: RegularBuilder.java:112-117
        this.weightRooms(this.multiConnections);
        Random.shuffle(this.multiConnections);

        // Remove duplicates while preserving order
        this.multiConnections = Array.from(new Set(this.multiConnections));

        // Shuffle again
        Random.shuffle(this.multiConnections);

        // Determine number of rooms on main path
        // Source: RegularBuilder.java:119
        let roomsOnMainPath = Math.floor(this.multiConnections.length * this.pathLength) +
            Random.chances(this.pathLenJitterChances);

        // Move rooms from multiConnections to mainPathRooms
        // Source: RegularBuilder.java:121-129
        while (roomsOnMainPath > 0 && this.multiConnections.length > 0) {
            const r = this.multiConnections.shift();
            // Simplified: assume sizeFactor() = 1 for all rooms
            roomsOnMainPath--;
            this.mainPathRooms.push(r);
        }
    }

    /**
     * Weight rooms (adds room multiple times for higher connection weight)
     * Source: RegularBuilder.java:134-141
     */
    weightRooms(rooms) {
        const toAdd = [];
        for (const r of rooms) {
            // Simplified: assume connectionWeight() = 1 for all rooms
            // In full implementation, StandardRoom has variable weights
        }
        rooms.push(...toAdd);
    }

    /**
     * Create branches from branchable rooms to roomsToBranch
     * Source: RegularBuilder.java:145-241
     */
    createBranches(rooms, branchable, roomsToBranch, connChances) {
        let i = 0;
        let angle;
        let tries;
        let curr;
        const connectingRoomsThisBranch = [];
        let failedBranchAttempts = 0;
        const connectionChances = connChances.slice();

        while (i < roomsToBranch.length) {
            // Fail if too many failed attempts
            // Source: RegularBuilder.java:157-159
            if (failedBranchAttempts > 100) {
                return false;
            }

            const r = roomsToBranch[i];

            connectingRoomsThisBranch.length = 0;

            // Pick random branchable room
            // Source: RegularBuilder.java:165-167
            curr = Random.element(branchable);

            // Determine number of connecting rooms
            // Source: RegularBuilder.java:169-174
            let connectingRooms = Random.chances(connectionChances);
            if (connectingRooms === -1) {
                connectionChances.splice(0, connectionChances.length, ...connChances);
                connectingRooms = Random.chances(connectionChances);
            }
            connectionChances[connectingRooms]--;

            // Place connecting tunnel rooms
            // Source: RegularBuilder.java:176-199
            for (let j = 0; j < connectingRooms; j++) {
                const t = new TunnelRoom();
                tries = 3;

                do {
                    angle = Builder.placeRoom(rooms, curr, t, this.randomBranchAngle(curr));
                    tries--;
                } while (angle === -1 && tries > 0);

                if (angle === -1) {
                    t.clearConnections();
                    for (const c of connectingRoomsThisBranch) {
                        c.clearConnections();
                        const idx = rooms.indexOf(c);
                        if (idx > -1) rooms.splice(idx, 1);
                    }
                    connectingRoomsThisBranch.length = 0;
                    break;
                } else {
                    connectingRoomsThisBranch.push(t);
                    rooms.push(t);
                }

                curr = t;
            }

            // Check if all connecting rooms were placed
            // Source: RegularBuilder.java:201-204
            if (connectingRoomsThisBranch.length !== connectingRooms) {
                failedBranchAttempts++;
                continue;
            }

            // Place the actual branch room
            // Source: RegularBuilder.java:206-222
            tries = 10;

            do {
                angle = Builder.placeRoom(rooms, curr, r, this.randomBranchAngle(curr));
                tries--;
            } while (angle === -1 && tries > 0);

            if (angle === -1) {
                r.clearConnections();
                for (const t of connectingRoomsThisBranch) {
                    t.clearConnections();
                    const idx = rooms.indexOf(t);
                    if (idx > -1) rooms.splice(idx, 1);
                }
                connectingRoomsThisBranch.length = 0;
                failedBranchAttempts++;
                continue;
            }

            // Add connecting rooms to branchable
            // Source: RegularBuilder.java:224-235
            for (let j = 0; j < connectingRoomsThisBranch.length; j++) {
                if (Random.Int_max(3) <= 1) {
                    branchable.push(connectingRoomsThisBranch[j]);
                }
            }
            if (r.maxConnections(0) > 1 && Random.Int_max(3) === 0) {
                branchable.push(r);
            }

            i++;
        }

        return true;
    }

    /**
     * Random branch angle
     * Source: RegularBuilder.java:243-245
     */
    randomBranchAngle(r) {
        return Random.Float() * 360;
    }
}
